[Merged by Bors] - Comprehensive CONTRIBUTING.md (#2040)
bevyengine/bevy: CONTRIBUTING.md

### Writing docs

Like every other large, rapidly developing open source library you've ever used, Bevy's documentation could stand to be improved.
This is incredibly valuable, easily distributed work, but requires a bit of guidance:

* inaccurate documentation is worse than no documentation: prioritize fixing broken docs
* Bevy is remarkable unstable: before tackling a new major documentation project, check in with the community about the stability of that feature and upcoming plans to save yourself heartache
* code documentation (doc examples and in the examples folder) is easier to maintain because the compiler will tell us when it breaks
* inline documentation should be fairly direct, minimal and technical
* the Bevy book is hosted on the `bevy-website` repo and targeted towards beginners who are just getting to know Bevy (and perhaps Rust!)
* RFCs are not documentation: they serve as a record of accepted decisions

[docs.rs](https://docs.rs/bevy) is built out of the last release's documentation, which is written right beside the adjacent code.
To view the current docs on `main` before you contribute, clone the `bevy` repo, and run `cargo doc --open`.

### Writing examples

> Most [examples in Bevy](https://github.com/bevyengine/bevy/tree/main/examples) aim to clearly demonstrate a single feature, group of closely related small features, or show how to accomplish a particular task (such as asset loading, creating a custom shader or how to test your app).

Reviewer: ```suggestion
Most [examples in Bevy](https://github.com/bevyengine/bevy/tree/main/examples) aim to clearly demonstrate a single feature, group of closely related small features, or show how to accomplish a particular task (such as asset loading, creating a custom shader or testing of your app).
```

In rare cases, creating new "game" examples is justified to demonstrate new features
that open a complex class of functionality in a way that's hard to showcase in isolation or requires additional integration testing.

Examples in Bevy should be:

1. **Working.** They must compile and run, and should fail in obvious ways.
2. **Clear.** They must use descriptive variable names, have reasonable code-quality, be formatted, and be appropriately commented.

> 3. **Relevant.** They should use game-relevant fluff and explain why what they're demonstrating is useful.

Reviewer: ```suggestion
3. **Relevant.** They should use game-relevant stuff and explain why what they're demonstrating is useful.
```

Author: Stuff isn't what I was going for, "fluff" is a jargon term to describe the aesthetics and theme as distinct from the mechanical details. E.g. you should show how resources work with a `Score` resource rather than `MyRes`.

We evidently need a clearer term for that though: any suggestions?

Reviewer: Oh wow, I was so sure that fluff was autocorrected from stuff that I didn't even stop to think that stuff does not fit here.

Would it be enough to just say "They should explain why what they're demonstrating is useful."

Author: Hmm. I think that plus "to a game developer" ought to do the trick. Lemme go change this.

Larger changes deserve a comment in the main thread, or a pull request to the original author's branch (but mention that you've made one).
When in doubt about a matter of architectural philosophy, refer back to **What we're trying to build** for guidance.

Once you're happy with the work and feel you're reasonably qualified to assess quality in this particular area, leave your `Approved` review on the PR so we can mark it as `ready-for-cart`.

There are a two main places you can check for new work to review:

1. Pull requests on `bevy` and the `bevy-website` repos.
2. [RFCs](https://github.com/bevyengine/rfcs), which need extensive thoughtful community input on their design.

Official focus areas and work done by @cart go through this review process as well.
Not even our project lead is exempt from reviews and RFCs!
By giving feedback on this work (and related supporting work), you can help us make sure our releases are both high-quality and timely.

Finally, if nothing brings you more satisfaction than seeing every last issue tagged and all resolved issues closed, feel free to message @cart for a Bevy org role to help us keep things tidy.

> As discussed in **How we're organized**, this is not intended to have a high bar.

Reviewer: Can we make references like `**How we're organized**` clickable links? The headers should get anchor tags in Github.

There are a few more of these references down the line.

### Contributing your own ideas

As discussed in **How we work together** Bevy is actively open to new ideas and serious contributions from outside community members.
If you're new to Bevy, here's the workflow we use:

> 1. Fork the `bevyengine/bevy` repository on GitHub, you'll need to create a GitHub account if you don't have one already.*

Reviewer: Did I find a rouge * here? ^^

Author: ```suggestion
1. Fork the `bevyengine/bevy` repository on GitHub, you'll need to create a GitHub account if you don't have one already.
```

2. Make your changes in a local clone of your fork, typically in its own new branch.

> 3. For a higher chance of CI passing the first time, consider locally running `cargo run -p ci`. You can run the commands manually:

Reviewer: Consider also adding info about markdownlint e.g. for changes in a readme. See https://github.com/bevyengine/bevy-website/pull/141

Author: Aha, I didn't think to check the open PRs for this info!

    1. `cargo fmt --all -- --check` (remove `--check` to let the command fix found problems)
    2. `cargo clippy --workspace --all-targets --all-features -- -D warnings -A clippy::type_complexity`
    3. `cargo test --all-targets --workspace`
4. Push your changes to your fork and open a Pull Request.
5. If you're a first time contributor to this repo, @cart [will need to manually trigger CI for your PR](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/). Feel free to ping him for this.
6. Respond to any CI failures or review feedback.

If you end up creating a new crate to the `bevy` repo:

1. Add a "Bevy Contributors <bevyengine@gmail.com>" entry in the Author field of `Cargo.toml`.

> 2. Add an MIT License to match the main `bevy` crate.

Reviewer: This is actually not the case at the moment. Do we want/need this? I guess it would make a lot of sense at least in cases like `bevy_ecs` where we explicitly say that they can be used standalone.

Author: Hmm, good question 🤔 I think it's clearer, but it depends on what shows up on e.g. docs.rs.

Author: Okay docs.rs correctly identifies the license as being inherited. Great, I'm removing this license.

### Writing examples

Most [examples in Bevy](https://github.com/bevyengine/bevy/tree/main/examples) aim to clearly demonstrate a single feature, group of closely related small features, or show how to accomplish a particular task (such as asset loading, creating a custom shader or testing of your app).
In rare cases, creating new "game" examples is justified to demonstrate new features
that open a complex class of functionality in a way that's hard to showcase in isolation or requires additional integration testing.

Examples in Bevy should be:

1. **Working.** They must compile and run, and should fail in obvious ways.
2. **Clear.** They must use descriptive variable names, have reasonable code-quality, be formatted, and be appropriately commented.
3. **Relevant.** They should use explain, through comments or variable names, how what they're demonstrating could be concretely useful to a game developer.
4. **Minimal.** They should be no larger or more complex than is needed to meet their other goals.This reduces maintenance burden and improves clarity when used as a reference.

When you add a new example, be sure to update `examples/README.md` with the new example.

> Use a generous sprinkling of keywords in your description: these are commonly used to locate a specific example using CTRL+F.

Reviewer 2: nit: "using CTRL+F" doesn't work on mac. of course every mac user is used to translate it directly but saying "when searching" achieve the same goal and is nicer imho

2. Make your changes in a local clone of your fork, typically in its own new branch.
3. For a higher chance of CI passing the first time, consider locally running `cargo run -p ci`. You can run the commands manually:
    1. `cargo fmt --all -- --check` (remove `--check` to let the command fix found problems)
    2. `cargo clippy --workspace --all-targets --all-features -- -D warnings -A clippy::type_complexity`
    3. `cargo test --all-targets --workspace`
4. When working with Markdown (`.md`) files, Bevy's CI will check markdown files like this one using [markdownlint](https://github.com/DavidAnson/markdownlint).
To locally lint your files using the same workflow as our CI:
   1. Install [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli).
   2. Run `markdownlint -f -c .github/linters/.markdown-lint.yml .` in the root directory of the Bevy project.
5. Push your changes to your fork and open a Pull Request.
6. If you're a first time contributor to this repo, @cart [will need to manually trigger CI for your PR](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/). Feel free to ping him for this.

> 7. Respond to any CI failures or review feedback.

Reviewer 2: ```suggestion
7. Respond to any CI failures or review feedback. CI failures must be fixed, review feedback should be answered but you are not obligated to follow them, use your best judgement.
```
not completely happy about my sentence here. I mean that while CI is mandatory, reviews from other members don't have the same weight and people are free to follow them or not. they may even get contradictory reviews!

Author: I like the sentiment. Pushing a commit with better wording for you.

> 

Reviewer 2: ```suggestion
8. When your PR gets close to merge, @cart will review it and suggest changes. If those changes are minimal he may even apply them directly to speed up merging.

```

Author: Added with minor wording changes.

## Getting oriented

Bevy, like any from-scratch game engine, is a large project!
It can be a bit overwhelming to start, so here's the bird's-eye view.

The main [Bevy engine org](https://github.com/bevyengine) has 4 important repos:

1. [`bevy`](https://github.com/bevyengine/bevy) where the engine itself lives, and the bulk of development work occurs.
2. [`bevy-website`](https://github.com/bevyengine/bevy-website) where the [official website](https://bevyengine.org/), release notes and Bevy book are hosted, created using the Zola static site generator for Rust.
3. [`awesome-bevy`](https://github.com/bevyengine/awesome-bevy) is a central home for community content: tutorials, tools, templates, showcases and crates! Make a PR if you want to showcase your stuff there!
4. [`rfcs`](https://github.com/bevyengine/rfcs) a place for informal but detailed discussion and design work for elaborate features and revamps.

The `bevy` repo itself contains many smaller subcrates, each of which can be downloaded on their own, and freely replaced to enable a modular architecture.

> Of particular interest, [`bevy_ecs`] is fully functional as a stand-alone ECS, which can be very valuable if you're looking to integrate it with other game engines or use it for non-game executables.
> 
> [`bevy_app`] and [`bevy_tasks`] are also worth calling out separately: the former serves as a nice framework to handle various data-piping needs, while the latter is our own lightweight custom async library.

Reviewer 2: ```suggestion
Of particular interest, [`bevy_ecs`](./crates/bevy_ecs) is fully functional as a stand-alone ECS, which can be very valuable if you're looking to integrate it with other game engines or use it for non-game executables.

[`bevy_app`](./crates/bevy_app) and [`bevy_tasks`](./crates/bevy_tasks) are also worth calling out separately: the former serves as a nice framework to handle various data-piping needs, while the latter is our own lightweight custom async library.
```

The bar for trust here is low due to the janitorial nature of the role; feel free to message @cart on GitHub or Discord after you've made a few contributions if you'd like to help out.

## How we work together

Making a game engine is a huge project, but at the moment we only have one paid contributor, @cart (go [donate!](https://github.com/sponsors/cart)).
While we have *many* active contributors (welcome aboard!), herding all of these cats in a predictable way is challenging.

Bevy releases are intended to be spaced 6-8 weeks apart and tend to target one or two major features, led by @cart.
[Once those features are complete](https://github.com/bevyengine/bevy/blob/main/docs/release_checklist.md),
we work to fix any `high-impact` or `regression` tagged issues that we can,
write up our release notes and migration guide, and then announce the next Bevy version to the world!

You can see what we're planning by following along at the [Bevy roadmap](https://github.com/bevyengine/bevy/projects/1).
If you'd like an up-to-the-minute update on progress, feel free to ask on Discord.

But as you may have guessed, that's not *all* that happens.
Simple changes have a simple process:

> 1. A community member creates an issue or opens a pull request to fix an issue or add simple functionality.

Reviewer 2: ```suggestion
1. A community member (that's you!) creates an issue or opens a pull request to fix an issue or add simple functionality.
```

> 2. Other communities review and comment on the work in an ad-hoc fashion.

Reviewer 2: ```suggestion
2. Other community members review and comment on the work in an ad-hoc fashion.
```

Some of the best ways to do this are:

* answering questions on [Discord](https://discord.gg/bevy), [GitHub Discussions](https://github.com/bevyengine/bevy/discussions), [reddit](https://www.reddit.com/r/bevy) or [StackOverflow](https://stackoverflow.com/questions/tagged/bevy).
* writing tutorials, guides and other informal documentation and sharing it on [awesome-bevy](https://github.com/bevyengine/awesome-bevy)
* streaming, writing blog posts about creating your game or creating videos. Share these in the `#devlogs` channel on Discord!

### Writing plugins

If you're interested in contributing to the community ecosystem in a way that doesn't make sense as part of the core engine,
feel free to write a plugin or crate for Bevy!

Any useful, reusable and non-trivial pieces of functionality that you've made serves as a good candidate for a plugin.
If it's closer to a snippet or design pattern, you may want to share it with the community on Discord, Reddit or GitHub Discussions instead.

Check out our [plugin guidelines](https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md) for helpful tips and norms.

### Fixing bugs

> Bugs in Bevy (or our website / books) are filed on the issue tracker corresponding to where they're found [], using the `bug` label.

Reviewer 2: ```suggestion
Bugs in Bevy (or our website / books) are filed on the issue tracker corresponding to where they're found [], using the [`bug`](https://github.com/bevyengine/bevy/issues?q=is%3Aissue+is%3Aopen+label%3Abug) label.
```

> Bugs in Bevy (or our website / books) are filed on the issue tracker corresponding to where they're found [], using the `bug` label.

Reviewer 2: what is the `[]`?

Author: Just a typo.

> If you're looking for an easy place to start, take a look at the `good-first-issue` tag, and feel free to ask questions on the issue in question or on Discord.

Reviewer 2: ```suggestion
If you're looking for an easy place to start, take a look at the [`good-first-issue`](https://github.com/bevyengine/bevy/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) tag, and feel free to ask questions on the issue in question or on Discord.
```

[docs.rs](https://docs.rs/bevy) is built out of the last release's documentation, which is written right beside the adjacent code.
To view the current docs on `main` before you contribute, clone the `bevy` repo, and run `cargo doc --open`.

### Writing examples

Most [examples in Bevy](https://github.com/bevyengine/bevy/tree/main/examples) aim to clearly demonstrate a single feature, group of closely related small features, or show how to accomplish a particular task (such as asset loading, creating a custom shader or testing of your app).
In rare cases, creating new "game" examples is justified to demonstrate new features
that open a complex class of functionality in a way that's hard to showcase in isolation or requires additional integration testing.

Examples in Bevy should be:

1. **Working.** They must compile and run, and should fail in obvious ways.
2. **Clear.** They must use descriptive variable names, have reasonable code-quality, be formatted, and be appropriately commented.
3. **Relevant.** They should use explain, through comments or variable names, how what they're demonstrating could be concretely useful to a game developer.

> 4. **Minimal.** They should be no larger or more complex than is needed to meet their other goals.

Reviewer 2: point 4 didn't get its own line, not sure why

This reduces maintenance burden and improves clarity when used as a reference.

> When you add a new example, be sure to update `examples/README.md` with the new example.

Reviewer 2: and add it in the root Cargo.toml file

> 4. **Minimal.** They should be no larger or more complex than is needed to meet their other goals.This reduces maintenance burden and improves clarity when used as a reference.

Reviewer 2: ```suggestion
4. **Minimal.** They should be no larger or more complex than is needed to meet their other goals. This reduces maintenance burden and improves clarity when used as a reference.
```

It's a nice place to chat about priorities, ask quick questions and get to know the other contributors and users in a less formal setting.

Read on if you're looking for:

* the high-level design goals of Bevy
* conventions and informal practices we follow when developing Bevy
* general advice on good open source collaboration practices
* concrete ways you can help us, no matter your background or skill level

We're thrilled to have you along as we build!

## Getting oriented

Bevy, like any from-scratch game engine, is a large project!
It can be a bit overwhelming to start, so here's the bird's-eye view.

The main [Bevy engine org](https://github.com/bevyengine) has 4 important repos:

1. [`bevy`](https://github.com/bevyengine/bevy) where the engine itself lives, and the bulk of development work occurs.
2. [`bevy-website`](https://github.com/bevyengine/bevy-website) where the [official website](https://bevyengine.org/), release notes and Bevy book are hosted, created using the Zola static site generator for Rust.

> 3. [`awesome-bevy`](https://github.com/bevyengine/awesome-bevy) is a central home for community content: tutorials, tools, templates, showcases and crates! Make a PR if you want to showcase your stuff there!

Reviewer 3: "crates/plugins (see below)". I'm not sure if my suggestion of adding plugins with a note to see below would be helpful, or unnecessary and confusing.

4. [`rfcs`](https://github.com/bevyengine/rfcs) a place for informal but detailed discussion and design work for elaborate features and revamps.

> The `bevy` repo itself contains many smaller subcrates, each of which can be downloaded on their own, and freely replaced to enable a modular architecture.

Reviewer 3: "can be downloaded on their own, and freely replaced" I'm not sure what would be more appropriate but the phrasing of 'freely replaced' doesn't seem quite right to me. I think it's an important point and the examples of bevy_ecs, bevy_app and bevy_tasks with the descriptions of what functionality they provide is helpful in clarifying it, but I don't fully grasp how each crate can be freely replaced to enable a modular architecture. Maybe an additional sentence of explanation would help?

Author: > The `bevy` repo itself contains many smaller subcrates, each of which can be downloaded on their own, and modularly replaced to allow you to pick and choose the parts of Bevy that you want to use.

Of particular interest, [`bevy_ecs`](./crates/bevy_ecs) is fully functional as a stand-alone ECS, which can be very valuable if you're looking to integrate it with other game engines or use it for non-game executables.

[`bevy_app`](./crates/bevy_app) and [`bevy_tasks`](./crates/bevy_tasks) are also worth calling out separately: the former serves as a nice framework to handle various data-piping needs, while the latter is our own lightweight custom async library.

## What we're trying to build

Bevy is intended as an **accessible**, **Rust-first**, **commercially-viable**, **free and open source** game engine.

> We prioritize:

Reviewer 3: "Bevy is still in it's early stages of development despite what we've accomplished so far. So, our priorities will change as the engine grows and matures. Below is a list of what we currently prioritize."

`We Prioritize:`

Author: Too verbose :p The intent is to convey values, not a roadmap. However it's good to hint that these aren't fixed forever.

* rapid experimentation over API stability

> * a consistent vision over a democratic decision making process

Reviewer 3: (see How We're Organized) (hyper-linked?)

> * accessible, lightweight workflows over bureaucratic standardization

Reviewer 3: I'm not clear on what "accessible, lightweight workflows" means

Author: "flexible processes"

> * a focus on our next goal over immediately integrating major new community-contributed features

Reviewer 3: "our next goal" perhaps "completing existing (ongoing? pre-existing?) work on big features over..."

* end-user ergonomics over implementation simplicity
* modularity over deep cross-crate integration
* supporting a thriving, easily integrated ecosystem over cramming every feature into the core engine

> * control over our code over saving work by reusing existing software

Reviewer 3: `saving` snipe `'saving'`

Author: Better to be sincere here IMO

* an ergonomic Rust workflow over a first-party scripting language
* thoughtful public interfaces over maximal configurability

> * welcoming contributions over insisting on existing skills and knowledge

Reviewer 3: ~~insisting on~~ `requiring`

> * an inclusive environment over avoiding conflict at any cost

Reviewer 3: maintaining an inclusive environment and a healthy, welcoming community over...

## How we're organized

> @cart is, for now, our singular Benevolent Dictator and project lead.

Reviewer 3: Benevolent Dictator For As Long As He Can Manage It (BDFALAHCMI?) ;)

Author: Redundant and too long :p

The bar for trust here is low due to the janitorial nature of the role; feel free to message @cart on GitHub or Discord after you've made a few contributions if you'd like to help out.

## How we work together

Making a game engine is a huge project, but at the moment we only have one paid contributor, @cart (go [donate!](https://github.com/sponsors/cart)).
While we have *many* active contributors (welcome aboard!), herding all of these cats in a predictable way is challenging.

Bevy releases are intended to be spaced 6-8 weeks apart and tend to target one or two major features, led by @cart.
[Once those features are complete](https://github.com/bevyengine/bevy/blob/main/docs/release_checklist.md),
we work to fix any `high-impact` or `regression` tagged issues that we can,
write up our release notes and migration guide, and then announce the next Bevy version to the world!

You can see what we're planning by following along at the [Bevy roadmap](https://github.com/bevyengine/bevy/projects/1).
If you'd like an up-to-the-minute update on progress, feel free to ask on Discord.

> But as you may have guessed, that's not *all* that happens.

Author: Clarified.

Simple changes have a simple process:

> 1. A community member (that's you!) creates an issue or opens a pull request to fix an issue or add simple functionality.

Author: Disagree due to the "issue" bit of the sentence.

By teaching others and helping them with their issues, you can learn the engine yourself and make our community better!

Some of the best ways to do this are:

* answering questions on [Discord](https://discord.gg/bevy), [GitHub Discussions](https://github.com/bevyengine/bevy/discussions), [reddit](https://www.reddit.com/r/bevy) or [StackOverflow](https://stackoverflow.com/questions/tagged/bevy).
* writing tutorials, guides and other informal documentation and sharing it on [awesome-bevy](https://github.com/bevyengine/awesome-bevy)
* streaming, writing blog posts about creating your game or creating videos. Share these in the `#devlogs` channel on Discord!

### Writing plugins

If you're interested in contributing to the community ecosystem in a way that doesn't make sense as part of the core engine,
feel free to write a plugin or crate for Bevy!

Any useful, reusable and non-trivial pieces of functionality that you've made serves as a good candidate for a plugin.
If it's closer to a snippet or design pattern, you may want to share it with the community on Discord, Reddit or GitHub Discussions instead.

> Check out our [plugin guidelines](https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md) for helpful tips and norms.

Author: Redundant to the guidelines before.

### Fixing bugs

Bugs in Bevy (or our website / books) are filed on the issue tracker corresponding to where they're found, using the [`bug`](https://github.com/bevyengine/bevy/issues?q=is%3Aissue+is%3Aopen+label%3Abug) label.

> If you're looking for an easy place to start, take a look at the [`good-first-issue`](https://github.com/bevyengine/bevy/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) tag, and feel free to ask questions on the issue in question or on Discord.

Reviewer 3: ~~the issue in question~~ that issue's thread or on Discord.

You don't need anyone's permission to try fixing a bug or adding a simple feature, but stating that you'd like to tackle an issue can be helpful to avoid duplicated work.

When you make a pull request that fixes an issue, include a line that says `Fixes #X` (or "Closes"), where `X` is the issue number.
This will cause the issue in question to be closed when your PR is merged.

General improvements to code quality are also welcome!
Bevy can always be safer, better tested, and more idiomatic.

### Writing docs

> Like every other large, rapidly developing open source library you've ever used, Bevy's documentation could stand to be improved.

Reviewer 3: rapidly-developing open-source library ~~you've ever used~~, Bevy's documentation can always use improvement.

This is incredibly valuable, easily distributed work, but requires a bit of guidance:

* inaccurate documentation is worse than no documentation: prioritize fixing broken docs

> * Bevy is remarkable unstable: before tackling a new major documentation project, check in with the community about the stability of that feature and upcoming plans to save yourself heartache

Reviewer 3: check ~~in~~ with the community on github or Discord about the...

* code documentation (doc examples and in the examples folder) is easier to maintain because the compiler will tell us when it breaks

> * inline documentation should be fairly direct, minimal and technical

Reviewer 3: ~~direct, minimal~~ short, to the point...

* the Bevy book is hosted on the `bevy-website` repo and targeted towards beginners who are just getting to know Bevy (and perhaps Rust!)

> * RFCs are not documentation: they serve as a record of accepted decisions

Reviewer 3: serve *only* as a record...

> [docs.rs](https://docs.rs/bevy) is built out of the last release's documentation, which is written right beside the adjacent code.

Reviewer 3: ~~built out of~~ built from

Reviewer 3: ~~written right beside the adjacent code.~~ written inline, above the corresponding code.

To view the current docs on `main` before you contribute, clone the `bevy` repo, and run `cargo doc --open`.

### Writing examples

> Most [examples in Bevy](https://github.com/bevyengine/bevy/tree/main/examples) aim to clearly demonstrate a single feature, group of closely related small features, or show how to accomplish a particular task (such as asset loading, creating a custom shader or testing of your app).

Reviewer 3: testing ~~of~~ your app).

> In rare cases, creating new "game" examples is justified to demonstrate new features

Reviewer 3: justified in order to

> that open a complex class of functionality in a way that's hard to showcase in isolation or requires additional integration testing.

Reviewer 3: ~~showcase~~ demonstrate

Examples in Bevy should be:

> 1. **Working:** They must compile and run, and should fail in obvious ways.

Reviewer 3: I think this could be expanded slightly, currently I don't understand the 'should fail in obvious ways' part.

> 2. **Clear:** They must use descriptive variable names, have reasonable code-quality, be formatted, and be appropriately commented.

Reviewer 3: ~~reasonable~~ good code-quality, be properly formatted...

> 3. **Relevant:** They should explain, through comments or variable names, how what they can be useful to a game developer.

Reviewer 3: ~~how what they~~ what they do, and how this can be useful...

4. **Minimal:** They should be no larger or more complex than is needed to meet their other goals.

When you add a new example, be sure to update `examples/README.md` with the new example and add it to the root `Cargo.toml` file.
Use a generous sprinkling of keywords in your description: these are commonly used to search for a specific example.

More complex demonstrations of functionality are also welcome, but for now belong in community tutorials or template games.

Check out [awesome-bevy](https://github.com/bevyengine/awesome-bevy) for a place to put your tutorials, tools, templates and crates!

### Reviewing others work

> With the sheer volume of activity in Bevy's community, reviewing others work and helping it improve is one of the most valuable things you can do.

Reviewer 3: ~~it improve~~ improve it

> You don't need to be an Elder Rustacean to be useful here: anyone can catch issues of missing tests, unclear docs, logic errors or so on.

Reviewer 3: ~~or~~ and so on.

If you have unusual skills (e.g. advanced familiarity with `unsafe` code, rendering knowledge or web development experience) or personal experience with a problem, try to prioritize those areas to ensure we can get appropriate expertise where we need it.

Focus on giving constructive, actionable feedback that results in real improvements to code quality or end-user experience.
If you don't understand why an approach was taken, please ask!

Small changes work well as comments or in-line suggestions on specific lines of codes.
Larger changes deserve a comment in the main thread, or a pull request to the original author's branch (but mention that you've made one).

> When in doubt about a matter of architectural philosophy, refer back to [*What we're trying to build*](https://github.com/bevy-engine/bevy/blob/main/CONTRIBUTING.md#what-were-trying-to-build) for guidance.

Reviewer 3: ~~but~~ make sure to mention...

Once you're happy with the work and feel you're reasonably qualified to assess quality in this particular area, leave your `Approved` review on the PR so we can mark it as `ready-for-cart`.

There are a two main places you can check for new work to review:

1. Pull requests on `bevy` and the `bevy-website` repos.
2. [RFCs](https://github.com/bevyengine/rfcs), which need extensive thoughtful community input on their design.

Official focus areas and work done by @cart go through this review process as well.
Not even our project lead is exempt from reviews and RFCs!
By giving feedback on this work (and related supporting work), you can help us make sure our releases are both high-quality and timely.

Finally, if nothing brings you more satisfaction than seeing every last issue tagged and all resolved issues closed, feel free to message @cart for a Bevy org role to help us keep things tidy.

> As discussed in [*How we're organized*](https://github.com/bevy-engine/bevy/blob/main/CONTRIBUTING.md#how-were-organized), this is not intended to have a high bar.

Reviewer 3: ~~this is not intended to have a high bar.~~ this janitorial role only requires at a minimum: demonstrating a willingness to contribute to Bevy by submitting a few pull requests to fix small issues.

### Contributing your own ideas

Bevy is actively open to new ideas and serious contributions from outside community members.
If you're new to Bevy, here's the workflow we use:

1. Fork the `bevyengine/bevy` repository on GitHub, you'll need to create a GitHub account if you don't have one already.
2. Make your changes in a local clone of your fork, typically in its own new branch.

>    1. Try to split your work into separate commits, each with a new purpose. Be particularly thoughtful of this when responding to reviews so we can see what's changed.

Reviewer 3: each with a ~~new~~ different purpose.

Reviewer 3: ~~thoughtful~~ mindful of this

Reviewer 3: so ~~we can~~ it's easy to see...

To locally lint your files using the same workflow as our CI:
   1. Install [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli).
   2. Run `markdownlint -f -c .github/linters/.markdown-lint.yml .` in the root directory of the Bevy project.
5. Push your changes to your fork and open a Pull Request.
6. If you're a first time contributor to this repo, @cart [will need to manually trigger CI for your PR](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/). Feel free to ping him for this.
7. Respond to any CI failures or review feedback. While CI failures must be fixed before we can merge your PR; you do not need to *agree* with all feedback from your reviews, merely acknowledge that it was given. If you cannot come to an agreement, leave the thread open and defer to @cart's final judgement.
8. When your PR is ready to merge, @cart will review it and suggest final changes. If those changes are minimal he may even apply them directly to speed up merging.

> If you end up creating a new crate to the `bevy` repo:

Reviewer 3: ~~creating~~ adding

1. Add a "Bevy Contributors <bevyengine@gmail.com>" entry in the Author field of `Cargo.toml`.

> 2. Add the new crate to the ./tools./publish.sh file.

Reviewer 4: I'm not sure about the stance of certain licenses, but IMO avoiding LGPL (v3) licenses would be best.
Since they require anyone who uses them to open-source their source code, and then would possible prevent some groups from using bevy.

And if so, we should leave a comment here about specific licenses?

Author: This will be for subcrates of Bevy itself, which are implicitly part of the same license.

For ecosystem crates, the plugin guidelines discusses our preference for non-permissive licenses.

Reviewer 3: I'm confused: I thought our preference was for permissive licences?

When contributing, please:

* try to loosely follow the workflow in [*How we work together*](https://github.com/bevy-engine/bevy/blob/main/CONTRIBUTING.md#how-we-work-together)
* explain what you're doing and why
* document new code with doc comments
* include clear, simple tests

> * add or improve the examples for new functionality

Reviewer 3: ~~for~~ with new...

* welcoming contributions over insisting on existing skills and knowledge
* fostering an inclusive environment over avoiding conflict at any cost

## How we're organized

@cart is, for now, our singular Benevolent Dictator and project lead.
He makes the final decision on both design and code changes within Bevy in order to ensure a coherent vision and consistent quality of code.
In practice, @cart serves as a shockingly accountable dictator: open to new ideas and to changing his mind in the face of compelling arguments or community consensus.
Check out the next section for details on how this plays out.

[Bevy org members](https://github.com/orgs/bevyengine/people) are contributors who help keep our repos tidy; they can label and close issues and PRs but do not have merge rights or any special authority within the community.
The bar for trust here is low due to the janitorial nature of the role; feel free to message @cart on GitHub or Discord after you've made a few contributions if you'd like to help out.

## How we work together

Making a game engine is a huge project, but at the moment we only have one paid contributor, @cart (go [donate!](https://github.com/sponsors/cart)).

> While we have *many* active contributors (welcome aboard!), herding all of these cats in a predictable way is challenging.

Reviewer 5: s/cats/birds :)

Bevy releases are intended to be spaced 6-8 weeks apart and tend to target one or two major features, led by @cart.
[Once those features are complete](https://github.com/bevyengine/bevy/blob/main/docs/release_checklist.md),
we work to fix any `high-impact` or `regression` tagged issues that we can,
write up our release notes and migration guide, and then announce the next Bevy version to the world!

You can see what we're planning by following along at the [Bevy roadmap](https://github.com/bevyengine/bevy/projects/1).
If you'd like an up-to-the-minute look at our progress on a specific project, feel free to ask on Discord.

### Simple changes

Simple changes have a simple process:

1. A community member (that's you!) creates an issue or opens a pull request to fix an issue or add simple functionality.

> 2. Other community members review and comment in an ad-hoc fashion. Active subject matter experts may be pinged into a thread.

Author: ```suggestion
2. Other community members review and comment in an ad-hoc fashion. Active subject matter experts may be pinged into a thread. If your PR has been quiet for a while and is ready for review, feel free to bring it up on Discord in an appropriate engine development channel.
```

3. Once they're content with the quality of the work (code quality, documentation, approach, need for functionality), individual reviewers leave "Approved" reviews.
4. After consensus has been reached (typically two approvals from the community or one for extremely simple changes) and CI passes, the `ready-for-cart` label is added.

> 5. When @cart has a good opportunity to pause from his implementation work, he performs a final code review on these pull requests and then presses the Big Merge Button (actually, he types `bors r+` to make sure we don't break `main` by accident).

Reviewer 6: ```suggestion
5. When @cart has a good opportunity to pause from his implementation work, he performs a final code review on these pull requests and then presses the Big Merge Button (actually, he runs [Bors](https://bors.tech/) by typing `bors r+` to make sure we don't break `main` by accident).
```

   1. Install [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli).
   2. Run `markdownlint -f -c .github/linters/.markdown-lint.yml .` in the root directory of the Bevy project.
5. Push your changes to your fork and open a Pull Request.
6. If you're a first time contributor to this repo, @cart [will need to manually trigger CI for your PR](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/). Feel free to ping him for this.
7. Respond to any CI failures or review feedback. While CI failures must be fixed before we can merge your PR; you do not need to *agree* with all feedback from your reviews, merely acknowledge that it was given. If you cannot come to an agreement, leave the thread open and defer to @cart's final judgement.
8. When your PR is ready to merge, @cart will review it and suggest final changes. If those changes are minimal he may even apply them directly to speed up merging.

If you end up adding a new official Bevy crate to the `bevy` repo:

1. Add a "Bevy Contributors <bevyengine@gmail.com>" entry in the Author field of `Cargo.toml`.

> 2. Add the new crate to the ./tools./publish.sh file.

Reviewer 6: ```suggestion
2. Add the new crate to the [./tools/publish.sh](./tools/publish.sh) file.
```

* [**`bevy_app`**](./crates/bevy_app): The api used to define Bevy Plugins and compose them together into Bevy Apps. 
* [**`bevy_tasks`**](./crates/bevy_tasks): Our light-weight async executor. This drives most async and parallel code in Bevy.
* [**`bevy_render`**](./crates/bevy_render): Our core renderer API. It handles interaction with the GPU, such as the creation of Meshes, Textures, and Shaders. It also exposes a modular Render Graph for composing render pipelines. All 2D and 3D render features are implemented on top of this crate.

## What we're trying to build

Bevy is a completely free and open source game engine built in Rust. It currently has the following design goals:

* **Capable**: Offer a complete 2D and 3D feature set
* **Simple**: Easy for newbies to pick up, but infinitely flexible for power users
* **Data Focused**: Data-oriented architecture using the Entity Component System paradigm
* **Modular**: Use only what you need. Replace what you don't like
* **Fast**: App logic should run quickly, and when possible, in parallel
* **Productive**: Changes should compile quickly ... waiting isn't fun

Bevy also currently has the following "development process" goals: 

> * **Rapid experimentation over API stability**: We need the freedom to experiment and iterate in order to build the best engine we can. This will change over 

Reviewer 7: Haha nice catch!

3. Have asked to join the Bevy Org. Reach out to @cart on Discord or [email us](bevyengine@gmail.com) if you are interested. Everyone is welcome to do this. We generally accept membership requests, so don't hesitate if you are interested!

Some Bevy Org members are also [Triage Team](https://github.com/orgs/bevyengine/teams/triage-team) members. These people can label and close issues and PRs but do not have merge rights or any special authority within the community. Existing Bevy Engine Org members can [automatically request membership](https://github.com/orgs/bevyengine/teams/triage-team/members). Once again, if you are interested don't hesitate to apply. We generally accept membership requests.

We heavily limit who has merge rights within the org because this requires a large amount of trust when it comes to ethics, technical ability, and ability to enforce consistent project direction. Currently, only @cart can merge every class of change. @mockersf is allowed to merge small "uncontroversial" changes, provided they have two approvals.

## How we work together

> Making a game engine is a huge project and facilitating collaboration is a lot of work. At the moment @cart is our only paid contributor, so [go sponsor him!](https://github.com/sponsors/cart).

Reviewer 8: ```suggestion
Making a game engine is a huge project and facilitating collaboration is a lot of work. At the moment @cart is our only paid contributor, so [go sponsor him!](https://github.com/sponsors/cart)
```

While we have *many* active contributors, herding all of these ~~cats~~ birds in a predictable way is challenging.

> Bevy releases are intended to be spaced 6-8 weeks apart and tend to target one or two major features, led by @cart.

Reviewer 8: This seems untrue in practice, is it an actual goal?

Reviewer 7: Yeah this got less true over time. I think we should just cut it.

[Once those features are complete](https://github.com/bevyengine/bevy/blob/main/docs/release_checklist.md), we work to fix any [P-High](https://github.com/bevyengine/bevy/issues?q=is%3Aopen+is%3Aissue+label%3AP-High), [P-Critical](https://github.com/bevyengine/bevy/issues?q=is%3Aopen+is%3Aissue+label%3AP-Critical) or [C-Regression](https://github.com/bevyengine/bevy/issues?q=is%3Aopen+is%3Aissue+label%3AC-Regression) labeled issues that we can, write up our release notes and migration guide, and then announce the next Bevy version to the world!

We track issues and pull requests that must be included in releases using [Milestones](https://github.com/bevyengine/bevy/milestones).

> You can see what we're planning by following along at the [Bevy Roadmap](https://github.com/bevyengine/bevy/projects/1).

Reviewer 8: The project board (and the pinned issues) are a little out of date I think. Linking to this here might give new users the wrong idea about how active Bevy development is. I remember another open source project which had a blog that was never updated. It made the project look dead to new users, even though it was very much alive.

Reviewer 7: Alice and I plan on giving the project board(s) some love prior to the next release. I think I'm fine with linking to it for now, but I agree with your point.
